core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c

```diff
@@ -1085,6 +1117,177 @@ posix_memalign_wrapper(wasm_exec_env_t exec_env, void **memptr, int32 align,
     return 0;
 }
 
+#if WASM_ENABLE_LIB_PTHREAD_SEMAPHORE != 0
+
+static int32
+sem_open_wrapper(wasm_exec_env_t exec_env, const char *name, int32 oflags)
+{
+    korp_sem *psem = NULL;
+    ThreadInfoNode *info_node = NULL;
+
+    /**
+     * For RTOS, global semaphore map is safe for share the same semaphore
+     * between task/pthread.
+     * For Unix like system, it's dedicated for multiple processes.
+     */
+
+    if ((info_node = bh_hash_map_find(sem_info_map, (void *)name))) {
+        return info_node->handle;
+    }
+
+    if (!(psem = os_sem_open(name, oflags))) {
+        goto fail1;
+    }
+
+    if (!(info_node = wasm_runtime_malloc(sizeof(ThreadInfoNode))))
+        goto fail2;
+
+    memset(info_node, 0, sizeof(ThreadInfoNode));
+    info_node->exec_env = exec_env;
+    info_node->handle = allocate_handle();
+    info_node->type = T_SEM;
+    info_node->u.sem = psem;
+    info_node->status = SEM_CREATED;
+
+    if (!append_thread_info_node(info_node)) {
+        wasm_runtime_free(info_node);
+        goto fail2;
+    }
+
+    if (!bh_hash_map_insert(sem_info_map, (void *)name, info_node))
+        goto fail3;
+
+    return info_node->handle;
+
+fail3:
+    delete_thread_info_node(info_node);
+fail2:
+    os_sem_close(psem);
+fail1:
+    return -1;
+}
+
+void
+sem_fetch_cb(void *key, void *value, void *user_data)
+{
+    (void)key;
+    SemCallbackArgs *args = user_data;
+    ThreadInfoNode *info_node = value;
+    if (args->handle == info_node->handle && info_node->status == SEM_CREATED) {
```

Comment:
A little confused about checking `args->handle == info_node->handle`, it means that the semaphore can only be used in the thread which created it. Other threads in the same wasm app can not use the semaphore?

Reply:
Samephore is identified by a string "name" in hashmap, but represented as a handle (int32) in wasm app, here we try to find  info node by handle.
Since the sem_info_map is global available, so all thread can use semaphore.

Reply:
OK, got it, thanks.